src-java/kilda-persistence-tinkerpop/src/main/java/org/openkilda/persistence/ferma/repositories/FermaFlowRepository.java

```diff
@@ -500,6 +500,24 @@ public long computeFlowsBandwidthSum(Set<String> flowIds) {
                         }));
     }
 
+    @Override
+    public Collection<Flow> findLoopedByFlowIdAndLoopSwitchId(String flowId, SwitchId switchId) {
+        return framedGraph().traverse(g -> {
+            GraphTraversal<Vertex, Vertex> vertexVertexGraphTraversal = g.V()
+                    .hasLabel(FlowFrame.FRAME_LABEL)
+                    .has(FlowFrame.LOOPED_PROPERTY, true);
+            if (flowId != null) {
+                vertexVertexGraphTraversal.has(FlowFrame.FLOW_ID_PROPERTY, flowId);
+            }
+            return vertexVertexGraphTraversal;
+        }).toListExplicit(FlowFrame.class).stream()
+                .map(Flow::new)
+                .filter(f -> switchId == null
```

Comment:
Can we move it to the DB level?

Reply:
From performance standpoint, it's better to refactor this method into 2 traverses with subsequent concatenation:

```
String switchIdAsStr = SwitchIdConverter.INSTANCE.toGraphProperty(switchId);

List<Flow> result = new ArrayList<>();
        
framedGraph().traverse(g -> g.V()
                .hasLabel(FlowFrame.FRAME_LABEL)
                .has(FlowFrame.SRC_SWITCH_ID_PROPERTY, switchIdAsStr)
                .has(FlowFrame.LOOPED_PROPERTY, true))
                .frameExplicit(FlowFrame.class)
                .forEachRemaining(frame -> result.add(new Flow(frame))));

framedGraph().traverse(g -> g.V()
                .hasLabel(FlowFrame.FRAME_LABEL)
                .has(FlowFrame.DST_SWITCH_ID_PROPERTY, switchIdAsStr)
               .has(FlowFrame.LOOPED_PROPERTY, true))
                .frameExplicit(FlowFrame.class)
                .forEachRemaining(frame -> result.add(new Flow(frame)));

return result;
```

In this approach the indexes are used instead of full scan.

Reply:
done